pyccl/halos/halo_model.py

```diff
@@ -0,0 +1,523 @@
+        self.lmass = np.linspace(self.precision['log10M_min'],
+                                 self.precision['log10M_max'],
+                                 self.precision['nlog10M'])
+        self.mass = 10.**self.lmass
+        self.m0 = self.mass[0]
+
+        if self.precision['integration_method_M'] not in ['spline',
+                                                          'simpson']:
+            raise NotImplementedError("Only \'simpson\' and 'spline' "
+                                      "supported as integration methods")
+        elif self.precision['integration_method_M'] == 'simpson':
+            from scipy.integrate import simps
+            self.integrator = simps
+        else:
+            self.integrator = self._integ_spline
+
+    def _integ_spline(self, fM, lM):
+        # Spline integrator
+        return _spline_integrate(lM, fM, lM[0], lM[-1])
+
+    def _hmf(self, hmf, cosmo, a, mass_def=None):
+        # Evaluates halo mass function at the correct mass
+        # values.
+        return hmf.get_mass_function(cosmo, self.mass,
+                                     a, mdef_other=mass_def)
+
+    def _hbf(self, hbf, cosmo, a, mass_def=None):
+        # Evaluates halo bias at the correct mass values.
+        return hbf.get_halo_bias(cosmo, self.mass,
+                                 a, mdef_other=mass_def)
+
+    def _I_0_1_from_arrays(self, hmf_a, hmf0, uk_a):
+        # Solves the integral:
+        #   \int dM n(M,a) * u(k,a|M)
+        # - hmf_a is an array of halo mass function values.
+        # - hmf0 is the value of the halo mass function at
+        #       a very low mass.
+        # - uk_a is the halo profile evaluated at the same
+        #       masses.
+        i1 = self.integrator(hmf_a[..., :] * uk_a,
+                             self.lmass)
+        return i1 + hmf0 * uk_a[..., 0]
+
+    def _I_1_1_from_arrays(self, hmf_a, hbf_a, hmf0, uk_a):
+        # Solves the integral:
+        #   \int n(M,a) * b(M,a) * u(k,a|M)
+        # - hmf_a is an array of halo mass function values.
+        # - hbf_a is an array of halo bias values.
+        # - hmf0 is the value of the halo mass function at
+        #       a very low mass.
+        # - uk_a is the halo profile evaluated at the same
+        #       masses.
+        i1 = self.integrator((hmf_a * hbf_a)[..., :] * uk_a,
+                             self.lmass)
+        return i1 + hmf0 * uk_a[..., 0]
+
+    def _check_massfunc(self, massfunc):
+        if not isinstance(massfunc, MassFunc):
+            raise TypeError("massfunc must be of type `MassFunc`")
+
+    def _check_hbias(self, hbias):
+        if not isinstance(hbias, HaloBias):
+            raise TypeError("hbias must be of type `HaloBias`")
+
+    def _check_prof(self, prof):
+        if not isinstance(prof, HaloProfile):
+            raise TypeError("prof must be of type `HaloProfile`")
+
+
+def halomod_mean_profile_1pt(cosmo, hmc, k, a, massfunc, prof,
```

Comment:
I would make this a method on HMCalculator and call it something like `I_1_0` or maybe `int_1_0` or w/e. Ditto for the other integral.

Reply:
These things already exist within `HMCalculator` as private methods. This is just me exposing them outside of it.

Reply:
OK, wait, let me see if I can do this in a cleaner way.

Reply:
Check now